freshmaker/lightblue.py

```diff
@@ -1579,5 +1579,239 @@ def _get_images_to_rebuild(image):
+            marked as directly affected
+        :param Iterable rpm_nvrs: the list of RPM NVRs with the fixes in the advisory
+        :param Iterable content_sets: the list of content sets that the RPMs in ``rpm_nvrs`` are
+            released in
+        """
+        for image_group in to_rebuild:
+            # Find the first index in image_group of an image that is not directly
+            # affected with parents that are also not directly affected
+            not_directly_affected_index = None
+            # Skip the first image in the group since it is always directly affected
+            for i, image in enumerate(image_group[1:]):
+                if image.nvr in directly_affected_nvrs:
+                    not_directly_affected_index = None
+                elif not_directly_affected_index is None:
+                    # We must add 1 since the first index was skipped in the for loop
+                    not_directly_affected_index = i + 1
+
+            # The image group does not end with one or more images that are not directly affected
+            if not_directly_affected_index is None:
+                continue
+
+            # Try replacing all the not directly affected images starting from the first one
+            for i in range(not_directly_affected_index, len(image_group)):
+                # On the first iteration, this is the last directly affected image in image_group
+                child_image = image_group[i - 1]
+                # Find the pullspec of the parent from the perspective of the child image
+                parent_pull_spec = None
+                for pull_spec, pull_spec_info in (child_image["parent_image_builds"] or {}).items():
+                    if pull_spec_info["id"] == child_image["parent_build_id"]:
+                        parent_pull_spec = pull_spec
+                        break
+                else:
+                    # The pullspec of the parent could not be determined
+                    continue
+
+                parent_image = image_group[i]
+                rpm_name_to_nvrs = {kobo.rpmlib.parse_nvr(nvr)["name"]: nvr for nvr in rpm_nvrs}
+                # Get the RPM NVRs that were fixed and apply to the parent image since
+                # get_fixed_published_image will ensure all those RPMs are present
+                parent_applicable_rpm_nvrs = set()
+                if not parent_image.get_rpms():
+                    log.warning("The parent image %s does not have an RPM manifest", parent_image.nvr)
+                    continue
+
+                for rpm in parent_image.get_rpms():
+                    if rpm_name_to_nvrs.get(rpm["name"]):
+                        parent_applicable_rpm_nvrs.add(rpm_name_to_nvrs[rpm["name"]])
+
+                fixed_published_image = self.get_fixed_published_image(
+                    parent_pull_spec, parent_applicable_rpm_nvrs, content_sets
```

Comment:
I'm wondering if it's a good idea to get latest published image by the `pull_spec`, sometimes the parent image can be published later and has different pull_spec with the data in brew (that was generated while building). For example, a parent image (no matter rebuilt due to dependency or affected directly) rebuilt in a freshmaker event always has internal osbs pull specs without registry.

Reply:
Good point. I suppose I can get the original `FROM` line from dist-git but that's hacky.

The only other option I thought of was searching for the latest published name and version that matches, however, it has a few drawbacks. For some images, they just build against the latest RHEL/UBI of a particular major version (e.g. rhel/rhel7:lastest). Upgrading to a higher release with the same name and version as the previously used parent image may not be correct. Additionally, it's possible that the image is built against a tag with a specific NVR (though not super likely), so upgrading this would not be correct. I don't think the current deduplication code takes this into account. Lastly, the highest NVR is not guaranteed to be the latest image for a tag.

What are your thoughts?

Reply:
> The only other option I thought of was searching for the latest published name and version that matches, however, it has a few drawbacks.

That's what I was thinking of and as you said, it's not guaranteed to be safe to just use the highest release image of the name-version. 

An interesting thing is some people want us to use latest published parent image instead of the original parent image, because the original parent can have some outdated configs (e.g. contest sets, CVP testing scripts), those layered images are not updated so frequently, to make Freshmaker rebuilds their images with the latest published parent, they need to rebuild a new image with the latest parent by manual and publish it. Currently we don't have a good solution to support this, because images can have different workflow or rules for building and updating there NVRs.